Tools/ci-run.sh

```diff
@@ -3,125 +3,164 @@
   python -m pip install -r doc-requirements.txt || exit 1
 else
-  STYLE_ARGS="--no-code-style";
+  STYLE_ARGS="--no-code-style"
 
   # Install more requirements
-  if [ -n "${PYTHON_VERSION##*-dev}" ]; then
-    if [ -z "${BACKEND##*cpp*}" ]; then
+  if [[ $PYTHON_VERSION == *"-dev" ]]; then
+    if [[ $BACKEND == *"cpp"* ]]; then
       echo "WARNING: Currently not installing pythran due to compatibility issues"
       # python -m pip install pythran==0.9.5 || exit 1
-    fi
-
-    if [ "$BACKEND" != "cpp" -a -n "${PYTHON_VERSION##pypy*}" -a -n "${PYTHON_VERSION##2*}" -a -n "${PYTHON_VERSION##*3.4}" ]; then
+    elif [[ $PYTHON_VERSION == "pypy"* && $PYTHON_VERSION == "2"* && $PYTHON_VERSION == *"3.4" ]]; then
       python -m pip install mypy || exit 1
     fi
   fi
 fi
 
 # Run tests
+echo "==== Running tests ===="
 ccache -s 2>/dev/null || true
 export PATH="/usr/lib/ccache:$PATH"
 
-if [ "$NO_CYTHON_COMPILE" != "1" -a -n "${PYTHON_VERSION##pypy*}" ]; then
-  CFLAGS="-O2 -ggdb -Wall -Wextra $(python -c 'import sys; print("-fno-strict-aliasing" if sys.version_info[0] == 2 else "")')" \
-  python setup.py build_ext -i \
-          $(if [ "$COVERAGE" == "1" ]; then echo " --cython-coverage"; fi) \
-          $(if [ "$CYTHON_COMPILE_ALL" == "1" ]; then echo " --cython-compile-all"; fi) \
-          $(python -c 'import sys; print("-j5" if sys.version_info >= (3,5) else "")') \
-      || exit 1
-  if [ -z "$COVERAGE" -a -z "$STACKLESS" -a -z "$LIMITED_API" -a -z "$CYTHON_COMPILE_ALL" -a -z "$EXTRA_CFLAGS" -a -n "${BACKEND//*cpp*}" ]; then
+if [[ "$OSTYPE" == "msys" ]]; then  # for MSVC cl
+  # /wd will disable warning
+  # 4711 warns that function `x` selected for automatic inline expansion
+  # 4127 warns that conditional expression is constant, should be fixed here https://github.com/cython/cython/pull/4317
+  # (off by default) 5045 warns that Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified
+  # (off by default) 4820 warns about the code in Python\3.9.6\x64\include ...
+  WARNARGS="/W4 /wd4711 /wd4127 /wd5045 /wd4820"
+  DEBUG_INFO="/Z7"
+  NO_OPTIMIZATION="-Od"
+else
+  WARNARGS="-Wall -Wextra"
+  DEBUG_INFO="-ggdb"
+  NO_OPTIMIZATION="-O0"
+fi
+
+if [[ $NO_CYTHON_COMPILE != "1" && $PYTHON_VERSION == "pypy"* ]]; then
+  if [[ $COVERAGE == "1" ]]; then
+    COVERAGE_ARGS="--cython-coverage"
+  fi
+  if [[ $CYTHON_COMPILE_ALL == "1" ]]; then
+    CYTHON_COMPILE_ALL_FLAGS=" --cython-compile-all"
+  fi
+  if [[ $PYTHON_SYS_VERSION > "3.5" || $PYTHON_SYS_VERSION == "3.5"* ]]; then
+    SETUP_ARGS="-j5"
+  fi
+  if [[ $PYTHON_SYS_VERSION == "2"* ]]; then
+    ALIASING="$-fno-strict-aliasing"
+  fi
+
+  CFLAGS="-O2 $DEBUG_INFO $WARNARGS $ALIASING" \
+    python setup.py build_ext -i \
+    $COVERAGE_ARGS $CYTHON_COMPILE_ALL_FLAGS $SETUP_ARGS || exit 1
+
+  if [[ $COVERAGE != "1" && $STACKLESS != "true" && -z $LIMITED_API && $CYTHON_COMPILE_ALL != "1" && -z $EXTRA_CFLAGS && $BACKEND != *"cpp"* ]]; then
```

Comment:
Please quote the variables. If one of them is empty, you'll get a syntax error otherwise.
(This also applies in various other places.)

Reply:
If one of them is empty (and there are already empty variables), we get `""` - an empty string

Reply:
Sorry, I meant "unset", not "empty".

Reply:
> Sorry, I meant "unset", not "empty".

Hmm .. that doesn't change anything :)

One way or another, the variable will have the value of an empty string.

How did you decide that we will get a syntax error?

Reply:
Ah, right. `[[ ... ]]` also avoids that issue. Good to know. It is an issue with single `[ ... ]` conditions.